File: solutions/ruby/kindergarten-garden/1/kindergarten_garden.rb
class Garden
  CHILDRENS = %w[Alice Bob Charlie David Eve Fred Ginny Harriet Ileana Joseph Kincaid Larry]
  PLANTS = {
    'V' => :violets,
    'R' => :radishes,
    'C' => :clover,
    'G' => :grass
  }

  CHILDRENS.each{|children|
    define_method(children.downcase) {
      @hash[children]
    }
  }


  def to_symbol(plant)
    PLANTS[plant] or plant
  end

  def initialize(garden)
    @hash = garden
    .split("\n")
    .map(&:chars)
    .transpose
    .each_slice(2)
    .to_a
    .map(&:transpose)
    .map(&:flatten)
    .zip(CHILDRENS)
    .each_with_object({}) {|(plants, children), hash|
      hash[children] = plants.map{|plant| to_symbol(plant)}
    }
  end

end

File: solutions/python/kindergarten-garden/1/kindergarten_garden.py
STUDENTS = [
    "Alice", "Bob", "Charlie", "David",
    "Eve", "Fred", "Ginny", "Harriet",
    "Ileana", "Joseph", "Kincaid", "Larry"

]

PLANTS = {
    'V': 'Violets',
    'R': 'Radishes',
    'C': 'Clover',
    'G': 'Grass'
}


class Garden:
    def __init__(self, diagram: str, students: list[str] = []):
        row_one, row_two = diagram.split('\n')
        self.students = sorted(students) or STUDENTS
        self.grouped_diagrams = [row_one[i:i+2] + row_two[i:i+2]
                                 for i in range(0, len(row_one), 2)]

    def plants(self, student: str):
        return [PLANTS[plant] for plant in self.grouped_diagrams[self.students.index(student)]]
